smallrye-reactive-messaging-aws-sns/src/main/java/io/smallrye/reactive/messaging/aws/sns/SnsConnector.java

```diff
@@ -0,0 +1,199 @@
+    private ExecutorService threadExecutor;
+    private Scheduler scheduler;
+
+    /**
+     * Method being invoked when CDI bean first created.
+     */
+    @Inject
+    public void initConnector() {
+        //Initialize vertx and threadExecutor
+        LOGGER.info("Initializing Connector");
+        if (instanceOfVertx != null && instanceOfVertx.isUnsatisfied()) {
+            internalVertxInstance = true;
+            this.vertx = Vertx.vertx();
+        } else if (instanceOfVertx != null) {
+            this.vertx = instanceOfVertx.get();
+        }
+        threadExecutor = Executors.newSingleThreadExecutor();
+        scheduler = Schedulers.single();
+    }
+
+    /**
+     * Its being invoked before bean being destroyed.
+     */
+    @PreDestroy
+    public void preDestroy() {
+        LOGGER.info("Destroying Connector");
+        if (internalVertxInstance) {
+            Optional.ofNullable(vertx).ifPresent(vertx -> vertx.close());
+        }
+        Optional.ofNullable(threadExecutor).ifPresent(exec -> exec.shutdown());
+        Optional.ofNullable(scheduler).ifPresent(sch -> sch.shutdown());
+    }
+
+    @Override
+    public SubscriberBuilder<? extends Message<?>, Void> getSubscriberBuilder(Config config) {
+
+        String topicName = getTopicName(config);
+        return ReactiveStreams.<Message<?>> builder().flatMapCompletionStage(message -> {
+            CompletionStage<Message<?>> cs = CompletableFuture.runAsync(() -> {
+                //send to sns
+                send(message, topicName);
+            }, threadExecutor).thenApply(x -> message);
+            return cs;
+        }).onError(t -> {
+            LOGGER.error("Error while subscribing to connector", t);
+        }).ignore();
+    }
+
+    /**
+     * Send message to SNS Topic.
+     * 
+     * @param msg Message to be sent
+     * @param topicName SNS topic name
+     */
+    private void send(Message<?> msg, String topicName) {
+
+        try (SnsAsyncClient snsClient = SnsClientManager.get().getAsyncClient()) {
+            //Prepare create topic request. if it is already created topicARN will be reutrned.
+            CreateTopicRequest topicCreationRequest = CreateTopicRequest.builder().name(topicName).build();
+            CompletableFuture<CreateTopicResponse> topicCreationResponse = snsClient.createTopic(topicCreationRequest);
+            String topicArn = topicCreationResponse.get().topicArn();
+            //Prepare publish message request to SNS topic
+            PublishRequest pr = PublishRequest
+                    .builder()
+                    .topicArn(topicArn)
+                    .message((String) msg.getPayload())
+                    .build();
+            CompletableFuture<PublishResponse> response = snsClient.publish(pr);
+            LOGGER.debug("Message ID {}", response.get().messageId());
+        } catch (InterruptedException | ExecutionException e) {
+            LOGGER.error("Async call interruption happened !!!", e);
```

Comment:
Wondering how they can happen now, with the completable future, as the failure will be reported on the future and not on this thread. What am I missing?

Reply:
It is a checked exception being thrown from AWS SDK Netty Async client. I need to handle it and at least log it.